monzo.js

```diff
@@ -1,133 +1,239 @@
+    delete successfulApplicants[arrayOfIDs[i]];
+    numberOfFailed++;
   }
-  console.log("\nOH FUCK, WE LOST " + subset.length + " PLEDGERS!" +
+
+  console.log("\nOH FUCK, WE LOST " + arrayOfIDs.length + " PLEDGERS!" +
     "\nNew number of chosen pledgers: " + Object.keys(successfulApplicants).length +
     "\nNew chosen pledges total: " + runningTotal
   );
+
+  saveVariablesToLocalStorage();
 }
 
+
 /**
  * If any pledges have been removed from the list of chosen pledges,
  * there will now be room for more to be chosen in their stead.
  * This function randomly selects as many additional pledges from
  * the initial list as needed to bring the total back up to as close
  * to the goal as possible
  */
-function repopulateChosenPledges() {
-  var maxPledgeValue = PLEDGE_VALUES[PLEDGE_VALUES.length - 1];
-  while (runningTotal <= GOAL - maxPledgeValue) {
-    var ids = Object.keys(allPledges);
-  	var randomId = ids[Math.floor(Math.random() * ids.length)];
-  	runningTotal += allPledges[randomId];
+monzo.repopulateSuccessfulApplicants = function () {
+  var maxPledgeValue = pledgeValues[pledgeValues.length - 1];
+  var ids = Object.keys(allPledges);
+
+  // Check that there are enough pledges in original set to repopulate
+  availablePledgesTotal = Object.keys(allPledges).reduce(function(sum, key) {
+    return sum + allPledges[key];
+  }, 0);
+  if (availablePledgesTotal < goal) {
+    alert(
+      "NOT ENOUGH PLEDGES TO REACH GOAL" +
+      "\nOnly " + availablePledgesTotal + " available towards goal of " + goal
+    );
+    throw Error("Not enough pledges to reach goal");
+  }
+
+  while (runningTotal <= goal - maxPledgeValue) {
+    var randomId = ids.splice(getRandomIndex(ids), 1)[0];
+    runningTotal += allPledges[randomId];
     successfulApplicants[randomId] = allPledges[randomId];
     delete allPledges[randomId];
   }
+
   console.log("\nNEW PLEDGERS SELECTED" +
     "\nNew number of chosen pledgers: " + Object.keys(successfulApplicants).length +
     "\nNew chosen pledges total: " + runningTotal +
     "\n"
   );
+
+  saveVariablesToLocalStorage();
 }
+
+
+/*---------------PRIVATE FUNCTIONS---------------*/
+
+function getCopyOfObject(oldDatasetObject) {
+  var newDatasetObject = {};
+  for (var i in oldDatasetObject) newDatasetObject[i] = oldDatasetObject[i];
+  return newDatasetObject;
+}
+
+function removeObjectSubset(subset, superset) {
+  for (var i in subset) delete superset[i];
+}
+
+function getRandomIndex(arr) {
+  return Math.floor(Math.random() * arr.length);
+}
+
+/*
+ * Returns an array of random keys, whose size is
+ * a specified fraction of that of the passed object
+ */
+function getRandomKeysFromObject(fractionOf, obj) {
+  var numberOf = Math.floor(Object.keys(obj).length * fractionOf);
+  var arr = [];
+  var keys = Object.keys(obj);
+  for (var i = 0; i < numberOf; i++)
+    arr.push(keys.splice(getRandomIndex(keys), 1)[0]);
+
+  return arr;
+}
+
+function saveVariablesToLocalStorage() {
```

Comment:
also does that mean i do not need getter/setter functions to retrieve local variables? because all my getter/setter functions will retrieve the stored variables instead?